### PR TITLE
Respect linearity when accumulating implicit params

### PR DESCRIPTION
**Pull request by MikeInnes**
All gradients need to be linear, i.e. consumed only once. Previously we allowed gradients to be consumed by the `accum_param` (tracking gradients of parameters given to the `Params` object) _and then_ continue backpropagating. Backpropagated gradients were always correct, but `accum_param` might be triggered more than once on the same object, leading to multiplied gradients. We now backpropagate `nothing` for a value instead.

### CODE REVIEW

File: src/lib/lib.jl [hunk 1]
```diff
@@ -46,10 +46,14 @@ end
 # backpropagated as values; these should be mutually exclusive options.
```
Comment (oxinabox) [inline src/lib/lib.jl:46]: Can this TODO be removed now?

Comment (MikeInnes) [reply src/lib/lib.jl:46]: Huh, guess I already thought of this issue at some point :D Yeah I think this is the fix though.

**Comment by cossio** [comment]
Can we merge this before the next tagged release? I think this is an important (silent!) bug. 
This probably also fixes https://github.com/FluxML/Zygote.jl/issues/691 (though I haven't tried yet).

**Comment by CarloLucibello** [comment]
tests are failing due to a return type inference problem. Can we fix this @MikeInnes ? Or just remove the test?

**Comment by MikeInnes** [comment]
Our type inference tests are pretty minimal and we should try to preserve them, since a performance issue with something like tuple indexing can easily become a big problem. I'm not sure why Julia doesn't figure it out in this case, but luckily it's easy to rewrite it in a way that can be inferred.

bors r+

**Comment by CarloLucibello** [comment]
@MikeInnes another CI error. Let's try to get this into the release

**Pull request merged**